File: kotlin/crypto-square/src/main/kotlin/CryptoSquare.kt
object CryptoSquare {

    fun ciphertext(plaintext: String): String {
        val normalized = plaintext.filter { it.isLetter() || it.isDigit() }.lowercase()
        val (r, c) = rectangle(normalized)

        return (0 until c).joinToString(" ") { i ->
            (0 until r).joinToString("") { j ->
                val index = j * c + i

                (if (index < normalized.length) {
                    normalized[index]
                } else {
                    ' '
                }).toString()
            }
        }
    }

    private fun rectangle(normalized: String): Pair<Int, Int> {
        val n = normalized.length
        var r = 1
        var c = n

        for (i in 1..n) {
            val rows = i
            val cols = (n + rows - 1) / rows

            if (cols >= rows && cols - rows <= 1) {
                r = rows
                c = cols
                break
            }
        }

        return Pair(r, c)
    }

}

File: csharp/crypto-square/CryptoSquare.cs
using System.Text;

public static class CryptoSquare
{
    public static string Ciphertext(string plaintext)
    {
        var normalized = string.Join("", plaintext.Where(c => char.IsLetter(c) || char.IsDigit(c))).ToLower();
        var (r, c) = Rectangle(normalized);

        var result = new List<string>();

        for (var i = 0; i < c; i++)
        {
            var sb = new StringBuilder();

            for (var j = 0; j < r; j++)
            {
                var index = j * c + i;
                sb.Append(index < normalized.Length ? normalized[index] : ' ');
            }

            result.Add(sb.ToString());
        }

        return string.Join(" ", result);
    }

    private static (int, int) Rectangle(string normalized)
    {
        var n = normalized.Length;
        var r = 1;
        var c = n;

        for (var i = 1; i <= n; i++)
        {
            var rows = i;
            var cols = (n + rows - 1) / rows;

            if (cols >= rows && cols - rows <= 1)
            {
                r = rows;
                c = cols;
                break;
            }
        }

        return (r, c);
    }
}
